Enterprise-CMCS/macpro-mdct-mcr
# Mdct 2274 - Kafka Integration

## Summary
Configures an integration with BigMac (Kafka) for downstream access to MCR reports

### Description
Borrowing logic from a few different sources for this.
SeaTools has an example of the new pattern for configuring and destroying topics [here](https://github.com/CMSgov/seatool-connectors/tree/master/src/services/topics). Sets up publishing to the topics in a similar manner as seds.

Breakdown:
* `services/topics` - Specifies topic names, and creates them on deploy.
* `services/app-api` - 
    * postKafkaData.ts - Configures the buckets and tables it will listen to for messages
    * kafka-source-lib.ts - Same lib for posting to kafka we use elsewhere, just with some extra logic to split dynamo messages and s3 messages, and handle them appropriately.
* `.github/workflows/destroy.yml` - Invokes the destroy function within `services/topics/` on branch destruction.

6 topics are created. Each report type gets one for its bucket and its dynamo table.
One advantage here is that downstream consumers can map explicitly which reports they want updates for.

### Related ticket
[MDCT-2274](https://qmacbis.atlassian.net/browse/MDCT-2274)

### How to test
~~Testing is going to require syncing up with someone on the Platform team who has access to topics. Editing a macpar report should publish a message to each of its topics. Maybe see if updates are appropriately made for most crud operations on a report on each topic.~~

I have verified with Mike that topics are created on this temp branch and are being written to.
We'll want to make sure **after this merges** that the temporary topics are cleaned up, and that the non-temporary branches also work.

Deployed branch here:
https://d1p1pbe8km4ds5.cloudfront.net/

### Important updates
Broker strings have been added to each aws account, but nothing should be required locally.

### Author checklist
